Issue 3502 - removing label propagation to build pods in build controller

soltysh opened the pull request:
Fixes #3502. I've added UPGRADE.md with the note that @smarterclayton proposed in the original issue, but I'm open to suggestion on renaming that file to something different. I couldn't make my mind between changelog and upgrade :wink:. I'm still working on it to prepare some more tests to make sure the controller is working properly with the old and new label.
@bparees @smarterclayton ptal

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -10,8 +10,12 @@ import (
+	DeprecatedBuildLabel = "build"
 	// BuildLabel is the key of a Pod label whose value is the Name of a Build which is run.
-	BuildLabel = "build"
+	BuildLabel = "io.openshift.build.name"
```
@liggitt do you think this is an OK change? I think we will make a life for our users harder as when you want to use label selector you will have to type the full namespace :-) it might look pretty ugly in CLI.

This format is a Docker images label format, but I don't we made any statement on using namespaces in `Labels` on our/k8s resources.

liggitt replied:
I would assume API label namespaces should match API annotation namespaces in style

liggitt replied:
Deployments also have a label used on pods related to a deployment: `openshift.io/deployer-pod-for.name`

mfojtik replied:
@liggitt correct, but the CLI issue is still valid:

`oc delete build all -l openshift.io/build/name=foo` vs. `oc delete build all -l build=foo`

soltysh replied:
Yeah, I messed this with Docker labels format, already fixed.

liggitt replied:
if people are frequently using build pod labels to manage builds pods from the cli, we're doing it wrong

smarterclayton replied:
Labels should be using "openshift.io/<>" - that is the Kubernetes
convention.  The Docker label format is only for image labels, not
Kubernetes labels.

On Tue, Jun 30, 2015 at 9:37 AM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/api/types.go
> <https://github.com/openshift/origin/pull/3508#discussion_r33572757>:
>
> >  	// BuildLabel is the key of a Pod label whose value is the Name of a Build which is run.
> > -	BuildLabel = "build"
> > +	BuildLabel = "io.openshift.build.name"
>
> @liggitt <https://github.com/liggitt> do you think this is an OK change?
> I think we will make a life for our users harder as when you want to use
> label selector you will have to type the full namespace :-) it might look
> pretty ugly in CLI.
>
> This format is a Docker images label format, but I don't we made any
> statement on using namespaces in Labels on our/k8s resources.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508/files#r33572757>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bparees commented on UPGRADE.md:
```diff
@@ -0,0 +1,7 @@
+This document describes future changes that will infect your current resources used inside of OpenShift.
```
s/infect/affect/

smarterclayton replied:
Add a header to this file - "Upgrading"

bparees commented on UPGRADE.md:
```diff
@@ -0,0 +1,7 @@
+This document describes future changes that will infect your current resources used inside of OpenShift.
+Each change contains description of the change and a date when that change will happen.
```
it's not a date, at best it's a release.

bparees commented on UPGRADE.md:
```diff
@@ -0,0 +1,7 @@
+
+* Build's pod label will be renamed from `build` to `io.openshift.build.name`.
+  All the newly created Builds will already have the new label, but old ones will
+  be recognized (inside the build controller) until post 3.1/1.2 version.
```
"old values will be recognized by OpenShift through the 3.1 release.  Support for these values may be removed in any major point release after 3.1 (e.g. 3.2)"

@smarterclayton please confirm that's the version guarantee we want to make here.

smarterclayton replied:
Let's say that we plan on removing this in the release that corresponds to
3.1.  The upgrade instructions for 3.1 will contain more info about making
this migration possible (probably by upgrading to the latest 3.0.x)

On Tue, Jun 30, 2015 at 10:09 AM, Ben Parees <notifications@github.com>
wrote:

> In UPGRADE.md
> <https://github.com/openshift/origin/pull/3508#discussion_r33576024>:
>
> > @@ -0,0 +1,7 @@
> > +This document describes future changes that will infect your current resources used inside of OpenShift.
> > +Each change contains description of the change and a date when that change will happen.
> > +
> > +* Build's pod label will be renamed from `build` to `io.openshift.build.name`.
> > +  All the newly created Builds will already have the new label, but old ones will
> > +  be recognized (inside the build controller) until post 3.1/1.2 version.
>
> "old values will be recognized by OpenShift through the 3.1 release.
> Support for these values may be removed in any major point release after
> 3.1 (e.g. 3.2)"
>
> @smarterclayton <https://github.com/smarterclayton> please confirm that's
> the version guarantee we want to make here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508/files#r33576024>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton replied:
Add an H2/H3 for the version this applies to (1.0.0) and then each item should be part of a list.  We should list as "OSE 3.1" and "Origin 1.x" when we know the Xs

soltysh replied:
What are H2/H3?

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -288,8 +288,12 @@ func (bc *BuildDeleteController) HandleBuildDeletion(build *buildapi.Build) erro
 	}
-	if pod.Labels[buildapi.BuildLabel] != build.Name {
-		glog.V(2).Infof("Not deleting pod %s/%s because the build label %s does not match the build name %s", pod.Namespace, podName, pod.Labels[buildapi.BuildLabel], build.Name)
+	if pod.Labels[buildapi.BuildLabel] != build.Name && pod.Labels[buildapi.DeprecatedBuildLabel] != build.Name {
```
should probably refactor this into a generic "getBuildLabel" method that returns string,bool(value,wasLabelFound)

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -357,31 +357,36 @@ type buildDeleteLW struct {
 	}
 	for _, pod := range podList.Items {
-		if len(pod.Labels[buildapi.BuildLabel]) == 0 {
+		if len(pod.Labels[buildapi.BuildLabel]) == 0 && len(pod.Labels[buildapi.DeprecatedBuildLabel]) == 0 {
```
and then use that new function here too.

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -357,31 +357,36 @@ type buildDeleteLW struct {
 		glog.V(5).Infof("Found build pod %s/%s", pod.Namespace, pod.Name)
 
-		build, err := lw.Client.Builds(pod.Namespace).Get(pod.Labels[buildapi.BuildLabel])
+		buildName := pod.Labels[buildapi.BuildLabel]
```
and here

mfojtik replied:
you can also check if the label exists in the map:

```go
buildName, found := pod.Labels[...]
if !found {
}
```

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -458,7 +463,7 @@ func (lw *buildPodDeleteLW) List() (runtime.Object, error) {
 			return nil, err
 		}
-		if (err != nil && kerrors.IsNotFound(err)) || pod.Labels[buildapi.BuildLabel] != build.Name {
+		if (err != nil && kerrors.IsNotFound(err)) || (pod.Labels[buildapi.BuildLabel] != build.Name && pod.Labels[buildapi.DeprecatedBuildLabel] != build.Name) {
```
and here

bparees commented on pkg/build/controller/strategy/util.go:
```diff
@@ -185,10 +185,5 @@ func getContainerVerbosity(containerEnv []kapi.EnvVar) (verbosity string) {
 
 // getPodLabels copies build labels and adds additional one with build name itself
```
comment is invalid now

bparees commented:
i'd call it DEPRECATED_FEATURES.md or DEPRECATED.md

assuming we're going to restrict it to deprecated items.  if we want it to be a more generic "here's stuff that has changed or may change in the future" then changelog makes sense to me.

"Change:  deprecated foo.  it may go away after release bar"

mfojtik commented on pkg/build/controller/factory/factory.go:
```diff
@@ -357,31 +357,36 @@ type buildDeleteLW struct {
 		}
 		if err != nil && kerrors.IsNotFound(err) {
 			build = nil
+
```
remove this extra line?

mfojtik commented on pkg/build/controller/factory/factory.go:
```diff
@@ -323,13 +323,13 @@ type podLW struct {
 // List lists all Pods that have a build label.
 func (lw *podLW) List() (runtime.Object, error) {
-	sel, _ := labels.Parse(buildapi.BuildLabel)
+	sel, _ := labels.Parse(buildapi.AllowedBuildLabels)
```
we should not mute the errors here

mfojtik commented on pkg/build/controller/factory/factory.go:
```diff
@@ -482,7 +487,7 @@ func (lw *buildPodDeleteLW) List() (runtime.Object, error) {
 // Watch watches all Pods that have a build label, for deletion
 func (lw *buildPodDeleteLW) Watch(resourceVersion string) (watch.Interface, error) {
-	sel, _ := labels.Parse(buildapi.BuildLabel)
+	sel, _ := labels.Parse(buildapi.AllowedBuildLabels)
```
same error muting here

smarterclayton commented:
UPGRADE.md is better.

On Tue, Jun 30, 2015 at 9:32 AM, Maciej Szulik <notifications@github.com>
wrote:

> Fixes #3502 <https://github.com/openshift/origin/issues/3502>. I've added
> UPGRADE.md with the note that @smarterclayton
> <https://github.com/smarterclayton> proposed in the original issue, but
> I'm open to suggestion on renaming that file to something different. I
> couldn't make my mind between changelog and upgrade [image: :wink:]. I'm
> still working on it to prepare some more tests to make sure the controller
> is working properly with the old and new label.
> @bparees <https://github.com/bparees> @smarterclayton
> <https://github.com/smarterclayton> ptal
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/openshift/origin/pull/3508
> Commit Summary
>
>    - Issue 3502 - removing label propagation to build pods in build
>
> File Changes
>
>    - *A* UPGRADE.md
>    <https://github.com/openshift/origin/pull/3508/files#diff-0> (7)
>    - *M* pkg/build/api/types.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-1> (6)
>    - *M* pkg/build/api/v1/types.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-2> (10)
>    - *M* pkg/build/api/v1beta3/types.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-3> (10)
>    - *M* pkg/build/controller/controller.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-4> (8)
>    - *M* pkg/build/controller/factory/factory.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-5> (21)
>    - *M* pkg/build/controller/strategy/custom_test.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-6> (7)
>    - *M* pkg/build/controller/strategy/docker_test.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-7> (7)
>    - *M* pkg/build/controller/strategy/sti_test.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-8> (7)
>    - *M* pkg/build/controller/strategy/util.go
>    <https://github.com/openshift/origin/pull/3508/files#diff-9> (7)
>
> Patch Links:
>
>    - https://github.com/openshift/origin/pull/3508.patch
>    - https://github.com/openshift/origin/pull/3508.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

soltysh commented:
I've addressed all the comments. I've also added quite some few tests to the build controller reaching ~90% test coverage :smiley: ptal. @smarterclayton is the current shape of UPGRADE.md exactly like you wanted it to look like?

soltysh commented:
[test] in the mean time

smarterclayton commented:
h2/h3 are header 2 and header 3 which in markdown are `##` and `###`

On Thu, Jul 2, 2015 at 8:28 AM, Maciej Szulik <notifications@github.com>
wrote:

> I've addressed all the comments. I've also added quite some few tests to
> the build controller reaching ~90% test coverage [image: :smiley:] ptal.
> @smarterclayton <https://github.com/smarterclayton> is the current shape
> of UPGRADE.md exactly like you wanted it to look like?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118013690>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bparees commented on UPGRADE.md:
```diff
@@ -0,0 +1,13 @@
+* Build's pod label will be renamed from `build` to `io.openshift.build.name`.
+  All the newly created Builds will already have the new label, old values will
+  be recognized by OpenShift through the 3.0.x release. Support for these values
+  will be removed in OSE 3.1 and Origin 1.x. See #3502.
```
the way this is structured it's a little hard to tell what version is impacted...  reading this i might think that "all the new created builds will already have the new label" applies to "3.0.0" which of course it doesn't.

I'd make it more clear what the current behavior is, what the transitional behavior will be,  and what the final behavior will be.

OSE 3.0.0
  currently all build pods have a label named build.  This label is being deprecated in favor of io.openshift.build.name in the X.Y.Z release (update this when we know what release the change is going into) at which point both labels will exist on build pods.  In version 3.1 the old labe (build) will be removed entirely.

soltysh replied:
Agreed, will update once I fix the integration tests.

smarterclayton replied:
These have to be tied to Origin version numbers first, with the
corresponding OSE number after.  When we don't know what either number will
be, we'll try to lock down.

On Thu, Jul 2, 2015 at 9:56 AM, Maciej Szulik <notifications@github.com>
wrote:

> In UPGRADE.md
> <https://github.com/openshift/origin/pull/3508#discussion_r33779740>:
>
> > @@ -0,0 +1,13 @@
> > +# Upgrading
> > +
> > +This document describes future changes that will affect your current resources used
> > +inside of OpenShift. Each change contains description of the change and information
> > +when that change will happen.
> > +
> > +
> > +## Origin 1.0.0 / OSE 3.0.0
> > +
> > +* Build's pod label will be renamed from `build` to `io.openshift.build.name`.
> > +  All the newly created Builds will already have the new label, old values will
> > +  be recognized by OpenShift through the 3.0.x release. Support for these values
> > +  will be removed in OSE 3.1 and Origin 1.x. See #3502.
>
> Agreed, will update once I fix the integration tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508/files#r33779740>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

soltysh commented:
@smarterclayton needs your wisdom with this. The problem with the test is in how we match new labels, I'm doing this `labels.Parse(buildapi.AllowedBuildLabels)` but that means we need both of these labels applied to Build's Pod, I was hoping there's an option to do or operation in `labels.Parse` but I cannot find it, this imho is bit too restrictive behavior. This PR is switching from using old label to the new one, I was hoping to apply only the new label starting from it but logic will be able to consume both, apparently there's no such option as of now. There are two possible solution to this situation:
1. easier - always apply both labels, in which case the migration will be a pain, because you will _have to_ change labels from an old to a new one during migration;
2. harder - implement or in the k8s selectors.
Unless I'm missing something here, which is also possible.

bparees commented:
applying both labels is definitely not a solution since we'd have to provide a migration tool that added the missing new label to all the existing objects (I suppose in theory we could add that logic to the converter and we might be covered, but it sounds terrible)

i think we need to add OR logic for KEYS to the upstream label selector if it's really not there.

smarterclayton commented:
You're going to have to do two queries.  Query new label, then old label.
No other solution is going to gracefully handle roll forward and rollback.

On Jul 4, 2015, at 9:15 PM, Ben Parees <notifications@github.com> wrote:

applying both labels is definitely not a solution since we'd have to
provide a migration tool that added the missing new label to all the
existing objects (I suppose in theory we could add that logic to the
converter and we might be covered, but it sounds terrible)

i think we need to add OR logic for KEYS to the upstream label selector if
it's really not there.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3508#issuecomment-118568233>.

bparees commented:
So double the amount of traffic pulled out of etcd? We really can't do an
OR selector?

Ben Parees | OpenShift
On Jul 5, 2015 10:26, "Clayton Coleman" <notifications@github.com> wrote:

> You're going to have to do two queries. Query new label, then old label.
> No other solution is going to gracefully handle roll forward and rollback.
>
> On Jul 4, 2015, at 9:15 PM, Ben Parees <notifications@github.com> wrote:
>
> applying both labels is definitely not a solution since we'd have to
> provide a migration tool that added the missing new label to all the
> existing objects (I suppose in theory we could add that logic to the
> converter and we might be covered, but it sounds terrible)
>
> i think we need to add OR logic for KEYS to the upstream label selector if
> it's really not there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118568233>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118626984>.
>

smarterclayton commented:
This is unlikely to be a performance issue.  OR selector us going to be
post Kube 1.0 because it's a big API change

On Jul 5, 2015, at 12:00 PM, Ben Parees <notifications@github.com> wrote:
Post

So double the amount of traffic pulled out of etcd? We really can't do an
OR selector?

Ben Parees | OpenShift
On Jul 5, 2015 10:26, "Clayton Coleman" <notifications@github.com> wrote:

> You're going to have to do two queries. Query new label, then old label.
> No other solution is going to gracefully handle roll forward and rollback.
>
> On Jul 4, 2015, at 9:15 PM, Ben Parees <notifications@github.com> wrote:
>
> applying both labels is definitely not a solution since we'd have to
> provide a migration tool that added the missing new label to all the
> existing objects (I suppose in theory we could add that logic to the
> converter and we might be covered, but it sounds terrible)
>
> i think we need to add OR logic for KEYS to the upstream label selector if
> it's really not there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118568233>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118626984>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3508#issuecomment-118635245>.

soltysh commented:
I was thinking about two queries and then combining the result as a viable
solution, but what about `Watch` function which also relies on that
selector? How should I combine that?
On Jul 5, 2015 9:09 PM, "Clayton Coleman" <notifications@github.com> wrote:

> This is unlikely to be a performance issue. OR selector us going to be
> post Kube 1.0 because it's a big API change
>
> On Jul 5, 2015, at 12:00 PM, Ben Parees <notifications@github.com> wrote:
> Post
>
> So double the amount of traffic pulled out of etcd? We really can't do an
> OR selector?
>
> Ben Parees | OpenShift
> On Jul 5, 2015 10:26, "Clayton Coleman" <notifications@github.com> wrote:
>
> > You're going to have to do two queries. Query new label, then old label.
> > No other solution is going to gracefully handle roll forward and
> rollback.
> >
> > On Jul 4, 2015, at 9:15 PM, Ben Parees <notifications@github.com> wrote:
> >
> > applying both labels is definitely not a solution since we'd have to
> > provide a migration tool that added the missing new label to all the
> > existing objects (I suppose in theory we could add that logic to the
> > converter and we might be covered, but it sounds terrible)
> >
> > i think we need to add OR logic for KEYS to the upstream label selector
> if
> > it's really not there.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/3508#issuecomment-118568233>.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/3508#issuecomment-118626984>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118635245>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118652908>.
>

smarterclayton commented:
Only watch the new label.  The resync will catch the old labels across an
upgrade.

On Jul 5, 2015, at 5:57 PM, Maciej Szulik <notifications@github.com> wrote:

I was thinking about two queries and then combining the result as a viable
solution, but what about `Watch` function which also relies on that
selector? How should I combine that?
On Jul 5, 2015 9:09 PM, "Clayton Coleman" <notifications@github.com> wrote:

> This is unlikely to be a performance issue. OR selector us going to be
> post Kube 1.0 because it's a big API change
>
> On Jul 5, 2015, at 12:00 PM, Ben Parees <notifications@github.com> wrote:
> Post
>
> So double the amount of traffic pulled out of etcd? We really can't do an
> OR selector?
>
> Ben Parees | OpenShift
> On Jul 5, 2015 10:26, "Clayton Coleman" <notifications@github.com> wrote:
>
> > You're going to have to do two queries. Query new label, then old label.
> > No other solution is going to gracefully handle roll forward and
> rollback.
> >
> > On Jul 4, 2015, at 9:15 PM, Ben Parees <notifications@github.com> wrote:
> >
> > applying both labels is definitely not a solution since we'd have to
> > provide a migration tool that added the missing new label to all the
> > existing objects (I suppose in theory we could add that logic to the
> > converter and we might be covered, but it sounds terrible)
> >
> > i think we need to add OR logic for KEYS to the upstream label selector
> if
> > it's really not there.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/3508#issuecomment-118568233>.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/3508#issuecomment-118626984>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118635245>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3508#issuecomment-118652908>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3508#issuecomment-118670200>.

soltysh commented:
Applied @smarterclayton to have double queries in all the lists and to use just the new label for `Watch`-es. Addressed @bparees comments to UPGRADE.md. ptal.

bparees commented on pkg/build/controller/controller_test.go:
```diff
@@ -607,3 +607,252 @@ func TestCancelBuild(t *testing.T) {
+	updateWasCalled := false
+	// finished build buildutil.IsBuildComplete should not be handled
+	build := mockBuild(buildapi.BuildStatusComplete, buildapi.BuildOutput{})
+	ctrl := mockBuildPodDeleteController(build, &customBuildUpdater{}, nil)
```
missing impl of customBuildUpdater to set updateWasCalled?

soltysh replied:
:blush: geez thx for noticing this

bparees commented on pkg/build/controller/controller_test.go:
```diff
@@ -607,3 +607,252 @@ func TestCancelBuild(t *testing.T) {
+	updateWasCalled := false
+	// erroneous builds should not be handled
+	build := mockBuild(buildapi.BuildStatusError, buildapi.BuildOutput{})
+	ctrl := mockBuildPodDeleteController(build, &customBuildUpdater{}, nil)
```
missing impl of customBuildUpdater to set updateWasCalled?

soltysh replied:
:blush: again thx.

bparees commented:
minor issue in your tests and then lgtm.

soltysh commented:
@bparees addressed your comments, @smarterclayton waiting for your sing-off.

smarterclayton commented:
@liggitt review w.r.t before after upgrade.

On Jul 6, 2015, at 4:09 PM, Maciej Szulik <notifications@github.com> wrote:

@bparees <https://github.com/bparees> addressed your comments,
@smarterclayton <https://github.com/smarterclayton> waiting for your
sing-off.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3508#issuecomment-118980999>.

bparees commented:
lgtm aside from the fact that it's going to cause me a merge conflict :(

liggitt commented on pkg/build/controller/factory/factory.go:
```diff
@@ -323,13 +323,35 @@ type podLW struct {
+	if err != nil {
+		return nil, err
+	}
+	listNew.Items = append(listNew.Items, listOld.Items...)
```
make sure the same build doesn't get in twice?

soltysh replied:
Should not happen, since old pods will have old label and new - the new one. The only time when the two would be there is a human intervention to add the second label.

smarterclayton replied:
Don't assume we won't run a migration or that won't happen - always defend
against that.

On Jul 7, 2015, at 7:23 AM, Maciej Szulik <notifications@github.com> wrote:

In pkg/build/controller/factory/factory.go
<https://github.com/openshift/origin/pull/3508#discussion_r34029858>:

> +		return nil, err
> +	}
> +	listNew, err := lw.client.Pods(kapi.NamespaceAll).List(sel, fields.Everything())
> +	if err != nil {
> +		return nil, err
> +	}
> +	// FIXME: get builds with old label - remove this when depracated label will be removed
> +	selOld, err := labels.Parse(buildapi.DeprecatedBuildLabel)
> +	if err != nil {
> +		return nil, err
> +	}
> +	listOld, err := lw.client.Pods(kapi.NamespaceAll).List(selOld, fields.Everything())
> +	if err != nil {
> +		return nil, err
> +	}
> +	listNew.Items = append(listNew.Items, listOld.Items...)

Should not happen, since old pods will have old label and new - the new
one. The only time when the two would be there is a human intervention to
add the second label.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3508/files#r34029858>.

liggitt commented on pkg/build/controller/factory/factory.go:
```diff
@@ -323,13 +323,35 @@ type podLW struct {
-	sel, _ := labels.Parse(buildapi.BuildLabel)
-	return lw.client.Pods(kapi.NamespaceAll).List(sel, fields.Everything())
+	// get builds with new label
+	sel, err := labels.Parse(buildapi.BuildLabel)
```
refactor the logic to a function that podLW.List() and buildDeleteLW.List() can both call?

liggitt commented on UPGRADE.md:
```diff
@@ -0,0 +1,14 @@
+## Origin 1.0.x / OSE 3.0.x
+
+* Currently all build pods have a label named `build`. This label is being deprecated
+  in favor of `io.openshift.build.name` in Origin 1.0.x / OSE 3.1.x at which point both
```
`openshift.io/build.name`

liggitt commented on UPGRADE.md:
```diff
@@ -0,0 +1,14 @@
+
+* Currently all build pods have a label named `build`. This label is being deprecated
+  in favor of `io.openshift.build.name` in Origin 1.0.x / OSE 3.1.x at which point both
+  labels will be supported. All the newely created builds will have just the new label.
```
typo: newly

liggitt commented on pkg/build/controller/factory/factory.go:
```diff
@@ -454,11 +491,15 @@ func (lw *buildPodDeleteLW) List() (runtime.Object, error) {
+				glog.V(4).Infof("Error getting pod for build %s/%s: %v", build.Namespace, build.Name, err)
+				return nil, err
+			} else {
+				pod = nil
```
continuing with a nil pod will cause a panic in `GetBuildLabel`

smarterclayton replied:
Did this comment get addressed?

soltysh replied:
Yes it did. Previously the call to `GetBuildLabel` was outside the if condition which might result in the situation @liggitt mentioned. Currently it is inside the else which is ok.

soltysh commented:
@liggitt addressed your comments.

soltysh commented:
@smarterclayton I've added function to `listPods` which handles duplicate pods. ptal

soltysh commented:
@bparees now you've caused me merge conflicts :stuck_out_tongue_winking_eye: but it was a small one. Rebased and waiting for final approval, @smarterclayton looking at you specifically.

smarterclayton commented:
You have build failures in travis

soltysh commented:
Right, sorry for that :blush: that was rebase leftovers from @bparees api changes. Should be ok now.

soltysh commented:
Jenkins is green, travis is failing int tests when with assets on timeout, can we increase that timeout for travis or somehow make the test on travis more reliable?

soltysh commented:
@smarterclayton, forgot to ping you here. Jenkins - green, travis is failing int tests with assets=true. Can we increase the timeout so that travis is more reliable?

liggitt commented:
FWIW, the timeout is on the go unit tests, not on the asset test (which passed)

soltysh commented:
I'm aware of that, I'm just saying that for some reason it's always timeouting when assets are tested as well.

smarterclayton commented:
LGTM [merge]

soltysh commented:
Strange error has happened, openshift-vagrant failed us upon syncing sources...
re-[merge]
